src/core/main/common/FreezeResolver.ts

```diff
@@ -0,0 +1,573 @@
+   *
+   * `null` if we never attempted to flush buffers.
+   */
+  private _lastFlushAttempt: {
+    /** Monotonically-raising timestamp at the time when we attempted the flush. */
+    timestamp: number;
+    /** Playback position at which the flush was performed, in seconds. */
+    position: number;
+  } | null;
+
+  /**
+   * If set to something else than `null`, this is the timestamp at the time the
+   * `FreezeResolver` started to consider its decipherability-linked un-freezing
+   * logic.
+   *
+   * This is used as a time of reference: after enough time was spent from that
+   * timestamp, the `FreezeResolver` will attempt supplementary unfreezing
+   * strategies.
+   *
+   * When the `FreezeResolver` is not considering those decipherability-related
+   * strategies for now, it is set to `null`.
+   */
+  private _decipherabilityFreezeStartingTimestamp: number | null;
+
+  constructor(segmentSinksStore: SegmentSinksStore) {
+    this._segmentSinksStore = segmentSinksStore;
+    this._decipherabilityFreezeStartingTimestamp = null;
+    this._ignoreFreezeUntil = null;
+    this._lastFlushAttempt = null;
+    this._lastSegmentInfo = {
+      audio: [],
+      video: [],
+    };
+  }
+
+  /**
+   * Check that playback is not freezing, and if it is, return a solution that
+   * should be attempted to unfreeze it.
+   *
+   * Returns `null` either when there's no freeze happening or if there's one
+   * but there's nothing we should do about it yet.
+   *
+   * Refer to the returned type's definition for more information.
+   *
+   * @param {Object} observation - The last playback observation produced, it
+   * has to be recent (just triggered for example).
+   * @returns {Object|null}
+   */
+  public onNewObservation(
+    observation: IFreezeResolverObservation,
+  ): IFreezeResolution | null {
+    const now = getMonotonicTimeStamp();
+    this._addPositionToHistory(observation, now);
+
+    if (this._ignoreFreezeUntil !== null && now < this._ignoreFreezeUntil) {
+      return null;
+    }
+    this._ignoreFreezeUntil = null;
+
+    const {
+      UNFREEZING_SEEK_DELAY,
+      UNFREEZING_DELTA_POSITION,
+      FREEZING_FLUSH_FAILURE_DELAY,
+    } = config.getCurrent();
+    const { readyState, rebuffering, freezing, fullyLoaded } = observation;
+
+    const freezingPosition = observation.position.getPolled();
+    const bufferGap = normalizeBufferGap(observation.bufferGap);
+
+    /** If set to `true`, we consider playback "frozen" */
+    const isFrozen =
+      freezing !== null ||
+      // When rebuffering, `freezing` might be not set as we're actively pausing
+      // playback. Yet, rebuffering occurences can also be abnormal, such as
+      // when enough buffer is constructed but with a low readyState (those are
+      // generally decryption issues).
+      (rebuffering !== null &&
+        readyState === 1 &&
+        (bufferGap >= MINIMUM_BUFFER_GAP_AT_READY_STATE_1_BEFORE_FREEZING ||
+          fullyLoaded));
```

Comment:
What if the `wantedBufferAhead` is set to a value lower than 6sec? The `bufferGap` should never be higher than 
`MINIMUM_BUFFER_GAP_AT_READY_STATE_1_BEFORE_FREEZING` in this case.

Reply:
Yeah :/

You think we should be including `wantedBufferAhead` there, reduce `MINIMUM_BUFFER_GAP_AT_READY_STATE_1_BEFORE_FREEZING`, or just consider this case as an edge case for which not handling freezing case would be acceptable?

We could also document in our API that a high enough `wantedBufferAhead` is preferable for our heuristics or adding logs.